### ISSUE TITLE
Buildalyzer fails with Invalid static method invocation syntax error on .NET Core

### ISSUE THREAD

**Issue by jmezach** [daveaglick/Buildalyzer#76]
We recently upgraded our static code analyzer to the latest version of Buildalyzer, including the MSBuild packages. Since then we're getting the following error:

```
Test Name:	RunAnalysisFromNetCoreProject
Test FullName:	Augurk.CSharpAnalyzer.UnitTest.AnalyzeCommandTest.RunAnalysisFromNetCoreProject
Test Source:	C:\Projects\open-source\Augurk\Augurk.CSharpAnalyzer\src\Augurk.CSharpAnalyzer.UnitTest\AnalyzeCommandTest.cs : line 26
Test Outcome:	Failed
Test Duration:	0:00:02,4173121

Result StackTrace:	
at Microsoft.Build.Shared.ProjectErrorUtilities.ThrowInvalidProject(String errorSubCategoryResourceName, IElementLocation elementLocation, String resourceName, Object[] args) in /_/src/Shared/ProjectErrorUtilities.cs:line 412
   at Microsoft.Build.Evaluation.Expander`2.Function`1.Execute(Object objectInstance, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation) in /_/src/Build/Evaluation/Expander.cs:line 3281
   at Microsoft.Build.Evaluation.Expander`2.PropertyExpander`1.ExpandPropertyBody(String propertyBody, Object propertyValue, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties) in /_/src/Build/Evaluation/Expander.cs:line 1198
   at Microsoft.Build.Evaluation.Expander`2.PropertyExpander`1.ExpandPropertiesLeaveTypedAndEscaped(String expression, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties) in /_/src/Build/Evaluation/Expander.cs:line 1039
   at Microsoft.Build.Evaluation.Expander`2.PropertyExpander`1.ExpandPropertiesLeaveEscaped(String expression, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties) in /_/src/Build/Evaluation/Expander.cs:line 896
   at Microsoft.Build.Evaluation.Expander`2.ExpandIntoStringLeaveEscaped(String expression, ExpanderOptions options, IElementLocation elementLocation) in /_/src/Build/Evaluation/Expander.cs:line 265
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluatePropertyElement(ProjectPropertyElement propertyElement) in /_/src/Build/Evaluation/Evaluator.cs:line 1446
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluatePropertyGroupElement(ProjectPropertyGroupElement propertyGroupElement) in /_/src/Build/Evaluation/Evaluator.cs:line 1087
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) in /_/src/Build/Evaluation/Evaluator.cs:line 954
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) in /_/src/Build/Evaluation/Evaluator.cs:line 1825
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) in /_/src/Build/Evaluation/Evaluator.cs:line 997
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) in /_/src/Build/Evaluation/Evaluator.cs:line 1825
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) in /_/src/Build/Evaluation/Evaluator.cs:line 997
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) in /_/src/Build/Evaluation/Evaluator.cs:line 1825
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) in /_/src/Build/Evaluation/Evaluator.cs:line 997
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) in /_/src/Build/Evaluation/Evaluator.cs:line 1825
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) in /_/src/Build/Evaluation/Evaluator.cs:line 997
   at Microsoft.Build.Evaluation.Evaluator`4.Evaluate(ILoggingService loggingService, BuildEventContext buildEventContext) in /_/src/Build/Evaluation/Evaluator.cs:line 726
   at Microsoft.Build.Evaluation.Project.Reevaluate(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings) in /_/src/Build/Definition/Project.cs:line 2752
   at Microsoft.Build.Evaluation.Project.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings) in /_/src/Build/Definition/Project.cs:line 2714
   at Microsoft.Build.Evaluation.Project.Initialize(IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext) in /_/src/Build/Definition/Project.cs:line 2822
   at Buildalyzer.ProjectAnalyzer.Load(String targetFramework, BuildEnvironment buildEnvironment)
   at Buildalyzer.ProjectAnalyzer.BuildTargets(BuildEnvironment buildEnvironment, String targetFramework, String[] targetsToBuild)
   at Buildalyzer.ProjectAnalyzer.Build(String targetFramework, BuildEnvironment buildEnvironment)
   at Buildalyzer.Workspaces.ProjectAnalyzerExtensions.AddToWorkspace(ProjectAnalyzer analyzer, Workspace workspace, Boolean addProjectReferences)
   at Augurk.CSharpAnalyzer.UnitTest.AnalyzeCommandTest.GetWorkspace(String solutionPath) in C:\Projects\open-source\Augurk\Augurk.CSharpAnalyzer\src\Augurk.CSharpAnalyzer.UnitTest\AnalyzeCommandTest.cs:line 49
   at Augurk.CSharpAnalyzer.Commands.AnalyzeCommand.Analyze(AnalyzeOptions options)
   at Augurk.CSharpAnalyzer.UnitTest.AnalyzeCommandTest.RunAnalysisFromNetCoreProject() in C:\Projects\open-source\Augurk\Augurk.CSharpAnalyzer\src\Augurk.CSharpAnalyzer.UnitTest\AnalyzeCommandTest.cs:line 38
Result Message:	
Test method Augurk.CSharpAnalyzer.UnitTest.AnalyzeCommandTest.RunAnalysisFromNetCoreProject threw exception: 
Microsoft.Build.Exceptions.InvalidProjectFileException: Invalid static method invocation syntax: "[Microsoft.Build.Utilities.ToolLocationHelper]::GetPathToStandardLibraries($(TargetFrameworkIdentifier), $(TargetFrameworkVersion), $(TargetFrameworkProfile), $(PlatformTarget), $(TargetFrameworkRootPath), $(TargetFrameworkFallbackSearchPaths))". Microsoft.Build.Utilities.ToolLocationHelper.GetPathToStandardLibraries Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)).   C:\Program Files\dotnet\sdk\2.1.401\Microsoft.Common.CurrentVersion.targets
```

This seems to be related to [this](https://developercommunity.visualstudio.com/content/problem/311136/update-to-1580-projects-wont-load.html) developer community issue, which suggests removing MSBuild assemblies from the GAC. I've checked on my machine, but only the MSBuild 14.0 assemblies are in the GAC (which as far as I know shouldn't be used since the assembly version doesn't match).

Interestingly we're seeing the same issue when running the tool on a Linux machine. The issue linked above also links to this [document](https://docs.microsoft.com/en-us/visualstudio/msbuild/updating-an-existing-application?view=vs-2017) suggesting to use the Microsoft.Build.Locator package, but that package only supports .NET Framework.

Any ideas what could be wrong here?

**Comment by daveaglick** [comment]
Not sure yet, but I'm actively troubleshooting issues when the latest .NET Core SDK is installed and the latest Buildalyzer is used. There's definitely an incompatibility somewhere - integration tests that work locally for me are failing on CI with very similar environments. Your problems _might_ be related and look similar to what I've been seeing regarding invalid project exceptions. Once I get to the bottom of it and release a new version lets see if that helps with your situation too.

**Comment by johtela** [comment]
Stumbled into the same issue when building a project targeting `net471`. I think the issue appeared when I updated to the 15.8.1 version of Visual Studio. With the previous version of VS Buildalyzer worked ok with the same project.

Here is the exception log, if it helps.
```
Microsoft.Build.Exceptions.InvalidProjectFileException
  HResult=0x80131500
  Message=Invalid static method invocation syntax: "[Microsoft.Build.Utilities.ToolLocationHelper]::GetPathToStandardLibraries($(TargetFrameworkIdentifier), $(TargetFrameworkVersion), $(TargetFrameworkProfile), $(PlatformTarget), $(TargetFrameworkRootPath), $(TargetFrameworkFallbackSearchPaths))". Microsoft.Build.Utilities.ToolLocationHelper.GetPathToStandardLibraries Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)).   C:\Program Files\dotnet\sdk\2.1.400\Microsoft.Common.CurrentVersion.targets
  Source=Microsoft.Build
  StackTrace:
   at Microsoft.Build.Shared.ProjectErrorUtilities.ThrowInvalidProject(String errorSubCategoryResourceName, IElementLocation elementLocation, String resourceName, Object[] args)
   at Microsoft.Build.Evaluation.Expander`2.Function`1.Execute(Object objectInstance, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation)
   at Microsoft.Build.Evaluation.Expander`2.PropertyExpander`1.ExpandPropertyBody(String propertyBody, Object propertyValue, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties)
   at Microsoft.Build.Evaluation.Expander`2.PropertyExpander`1.ExpandPropertiesLeaveTypedAndEscaped(String expression, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties)
   at Microsoft.Build.Evaluation.Expander`2.PropertyExpander`1.ExpandPropertiesLeaveEscaped(String expression, IPropertyProvider`1 properties, ExpanderOptions options, IElementLocation elementLocation, UsedUninitializedProperties usedUninitializedProperties)
   at Microsoft.Build.Evaluation.Expander`2.ExpandIntoStringLeaveEscaped(String expression, ExpanderOptions options, IElementLocation elementLocation)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluatePropertyElement(ProjectPropertyElement propertyElement)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluatePropertyGroupElement(ProjectPropertyGroupElement propertyGroupElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement)
   at Microsoft.Build.Evaluation.Evaluator`4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport)
   at Microsoft.Build.Evaluation.Evaluator`4.Evaluate(ILoggingService loggingService, BuildEventContext buildEventContext)
   at Microsoft.Build.Evaluation.Project.Reevaluate(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings)
   at Microsoft.Build.Evaluation.Project.ReevaluateIfNecessary(ILoggingService loggingServiceForEvaluation, ProjectLoadSettings loadSettings)
   at Microsoft.Build.Evaluation.Project.Initialize(IDictionary`2 globalProperties, String toolsVersion, String subToolsetVersion, ProjectLoadSettings loadSettings, EvaluationContext evaluationContext)
   at Buildalyzer.ProjectAnalyzer.Load(String targetFramework, BuildEnvironment buildEnvironment)
   at Buildalyzer.ProjectAnalyzer.BuildTargets(BuildEnvironment buildEnvironment, String targetFramework, String[] targetsToBuild)
   at Buildalyzer.ProjectAnalyzer.Build(String targetFramework, BuildEnvironment buildEnvironment)
   at LiterateCS.Weaver.BuildSolution() in C:\Users\johteto\Source\Repos\LiterateCS\LiterateCS\Weaver.cs:line 210
```

**Comment by daveaglick** [comment]
I just published Buildalyzer 2.0 which runs MSBuild out-of-process (as opposed to the old way of running it via MSBuild APIs). In general, if you can build it Buildalyzer should now be able to build it. Given how drastically different this new technique is, I'm closing all existing bug reports.

Please try the new version and if you're still having problems, please open a new issue.

**Comment by johtela** [comment]
I can happily report that version 2.0.1 fixed the issue for me.

Great job, and big thanks!
